### ISSUE TITLE
Publish of WPF app fails for ARM64 - "There was no runtime pack"

### ISSUE THREAD

**Issue by clairernovotny** [dotnet/sdk#15972]
Using .NET 6 Preview 1, trying to publish a WPF app for ARM64 fails with

`1>There was no runtime pack for Microsoft.WindowsDesktop.App available for the specified RuntimeIdentifier 'win-arm64'.`

Repro:

- Clone https://github.com/NuGetPackageExplorer/NuGetPackageExplorer
- Checkout the `arm64-support` branch.
- `dotnet publish -c release .\PackageExplorer\NuGetPackageExplorer.csproj /p:PublishProfile=Properties\PublishProfiles\WinARM64.pubxml`

```
 pwsh> dotnet publish -c release .\PackageExplorer\NuGetPackageExplorer.csproj /p:PublishProfile=Properties\PublishProfiles\WinARM64.pubxml
Microsoft (R) Build Engine version 16.9.0-preview-21103-02+198f3f262 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  Restored D:\dev\NuGetPackageExplorer\PackageExplorer\NuGetPackageExplorer.csproj (in 801 ms).
  Restored D:\dev\NuGetPackageExplorer\PackageViewModel\PackageViewModel.csproj (in 801 ms).
  Restored D:\dev\NuGetPackageExplorer\Core\Core.csproj (in 1.07 sec).
  1 of 4 projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
C:\Program Files\dotnet\sdk\6.0.100-preview.1.21103.13\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(389,5): error NETSDK1082: There was no runtime pack for Microsoft.WindowsDesktop.App available for the specified RuntimeIdentifier 'win-arm64'. [D:\dev\NuGetPackageExplorer\Core\Core.csproj]
```

**Comment by clairernovotny** [comment]
This seems to happen if there's a `<RuntimeIdentifiers>win-x86;win-x64;win-arm64</RuntimeIdentifiers>` property set.

**Comment by clairernovotny** [comment]
If I remove the `RuntimeIdentifiers` property, it succeeds with `dotnet publish` but fails in VS with:

```
1>Assets file 'D:\dev\NuGetPackageExplorer\PackageExplorer\obj\project.assets.json' doesn't have a target for 'net6.0-windows10.0.19041/win-x64'. Ensure that restore has run and that you have included 'net6.0-windows10.0.19041' in the TargetFrameworks for your project. You may also need to include 'win-x64' in your project's RuntimeIdentifiers.
```

You can see this by opening the solution in VS, and on the PackageExplorer.Package project, choose "Publish -> Create App Packages" and follow the prompts for sideloading. It'll build and fail without the RuntimeIdentifiers property (in the Directory.Build.props).

**Comment by marcpopMSFT** [comment]
@vijayrkn for the VS experience.  @dsplaisted for the RIDs question.

**Comment by clairernovotny** [comment]
For NPE, the `RuntimeIdentifiers` property is set here: https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/arm64-support/Directory.Build.props#L26

Setting it at other levels lead to the errors around things missing from the assets file.

**Comment by clairernovotny** [comment]
In  the latest dogfood 16.10 I see this when I load up the `main` version of NPE:

`SkipGetTargetFrameworkProperties` is set to true in the `ProjectReference`.
https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/main/PackageExplorer.Package/PackageExplorer.Package.wapproj#L81

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134). Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-arm-aot. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-arm. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-arm64-aot. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-x64-aot. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-x64. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-x86-aot. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error	NU1201	Project NuGetPackageExplorer is not compatible with uap10.0.17134 (UAP,Version=v10.0.17134) / win10-x86. Project NuGetPackageExplorer supports: net6.0-windows10.0.19041 (.NETCoreApp,Version=v6.0)	PackageExplorer.Package	D:\dev\NuGetPackageExplorer\PackageExplorer.Package\PackageExplorer.Package.wapproj	1	
Error		Project '..\PackageExplorer\NuGetPackageExplorer.csproj' targets 'net6.0-windows10.0.19041'. It cannot be referenced by a project that targets '.NETFramework,Version=v4.5.1'.	PackageExplorer.Package	C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets	1720	
```

**Comment by wli3** [comment]
@sfoslund I see we have this test ItCantPublishArm64Wpf  https://github.com/dotnet/sdk/blob/9c5126233b3d8a4ef1b0d2489098f29ec13389aa/src/Tests/Microsoft.NET.Build.Tests/KnownRuntimePackTests.cs#L138 

So it is expected arm64 does not work with WPF?

**Comment by clairernovotny** [comment]
Publish does work though on the dotnet cli. It's the combination of build/publish/wapproj in VS that's broken.

**Comment by vijayrkn** [comment]
@clairernovotny - If you specify just one RID in the pubxml, does it work from VS?
`<RuntimeIdentifier>win-arm64</RuntimeIdentifier>`

Based on the above error message, VS is not looking in the correct restore directory.

**Comment by isaacrlevin** [comment]
Seeing this issue as well. No matter what I put in there, it fails for targets. I do not have ARM support enabled. Also I am not targeting .NET 6

https://github.com/isaacrlevin/presencelight/blob/main/src/DesktopClient/PresenceLight.Package/PresenceLight.Package.wapproj

**Comment by LanceMcCarthy** [comment]
Tl;dr
Using `msbuild my.sln` stopped working, I ended up fixing it by using  `msbuild myWpf.csproj` instead. 


## Problem Summary

I'm seeing the same, but not sure what the affected change is responsible. See the failed build output here https://github.com/LanceMcCarthy/MediaFileManager/runs/2231000877?check_suite_focus=true  

These are the only changes since my last successful release pipeline deployment:

* `Microsoft.Windows.CsWinRT` updated from 1.1.1 to 1.1.4
* `Microsoft.AppCenter.Crashes` updated from 4.1.0 to 4.2.0
* `Microsoft.AppCenter.Analytics` updated from 4.1.0 to 4.2.0
* VS2019 16.9 is now being used in the GitHub Actions runner

### Facts 

- I did not add any new target platforms to the WAP project. I was getting exactly the same errors as Claire.
- Project is a WPF with  `net5.0-windows10.0.19041.0` 
- You can [see the csproj file here](https://github.com/LanceMcCarthy/MediaFileManager/blob/main/src/MediaFileManager/MediaFileManager.Desktop/MediaFileManager.Desktop.csproj), it is very simple and straightforward.
- This is public project with public GH Actions builds https://github.com/LanceMcCarthy/MediaFileManager/ 
- The build only fails when attempting to use `msbuild my.sln` step, **not** the `msbuild PackageProject.wapproj` step. 


## Accidentally Fixed It

While I was preparing and organizing the info for this reply, I decided to try a change in the CD workflow so that it identically matches the approach I use in my CI-only workflow

This is the broken version, it used to work in December (when GH Actions used VS 2019 16.8)
```
run: msbuild src/MySolution.sln /t:Restore /p:Configuration=Release /p:RuntimeIdentifier=win-x86
```

Working verison
```
run: msbuild src/WpfProject/MyWpfPrject.csproj /t:Restore /p:Configuration=Release /p:RuntimeIdentifier=win-x86
```

> Notice the difference between them is I am pointing msbuild directly at the WPF project's csproj instead of the sln. 

Here's a screenshot in case it helps:
![image](https://user-images.githubusercontent.com/3520532/113058451-23f28280-917c-11eb-9328-778b271abd99.png)

**Comment by vatsan-madhavan** [comment]
Seeing this issue as well.

I have a single RID in a pubxml file that changes with `$(Platfrom)` (something like `win-$(Platform)`), and I'm just using WinForms for settings support (`SettingsSingleFileGenerator`).

I don't get the sense that this problem turns on `RuntimeIdentifiers` vs. `RuntimeIdentifier`. 

- **sdk**: 6.0.100-preview.6.21355.2
- **tfm**: net5.0-windows10.0.17763.0
  - also repro's with net6.0-windows10.0.17763.0

**Comment by Samuel12321** [comment]
Having the same problem, is there any progress on a fix @vijayrkn @wli3 @marcpopMSFT

**Comment by sr55** [comment]
I ran into this issue as well however I've now managed to get past it. I'm using dotnet publish with a defined profile. 

In  my profile file, I have the following:

    <RuntimeIdentifier>win-arm64</RuntimeIdentifier>
    **<SelfContained>false</SelfContained>**                                 
    <PublishSingleFile>True</PublishSingleFile>
    **<PublishReadyToRun>false</PublishReadyToRun>**


The two highlighted in bold, have to be set to false.  Makes sense I guess but the error is very unintuitive. 

I also got caught out with one of my dependencies hooking a .NET Framework DLL (which is a no-go) but alas, different issue but something to watch out for too :)

Hopefully this is helpful to some.

**Comment by Samuel12321** [comment]
Thanks for the suggestion, unfortunately I need to publish self contained.

**Comment by sr55** [comment]
I have a feeling it would work if it was compiled on an ARM device, but not if your "cross compiling" on an x86 machine.

I suspect what's going on here, is that it's trying to find the ARM version of the SDK, and your machine I'm guessing has the x86/64 SDK.

**Comment by vatsan-madhavan** [comment]
This has suddenly started working for me with no changes from my side, with SelfContained=true, PublishReadyToRun=true, and still using SDK=5.0.302. 

@marcpopMSFT could it have had something to do with recent download problems related to CDN outage https://github.com/dotnet/core/issues/6509?

**Comment by marcpopMSFT** [comment]
That issue started on Friday so anyone having problems before then would not be from that issue.  Certainly anyone hitting it between the 23rd and 27th only could have been affected by the outage.

Per above, setting self-contained to false works around this as we only use the runtime packs for self-contained applications.  Given this is an old issue, I'm having trouble tracking if it's just the VS publish that's still broken or CLI publish as well.

**Comment by vatsan-madhavan** [comment]
Thanks @marcpopMSFT, that makes sense!

My original report on Jul 21st was CLI based (but I think I saw the same problem in VS as well). Unfortunately, my original report was based on observation from only one machine (my development machine). 

Today when I noticed that it's working, this was based on dotnet CLI observations. I was going to try out @sr55's idea of publishing `SelfContained=false`, but happened to build with `SelfContained=true`, `PublishReadyToRun=true` out of curiosity and noticed it built successfully. Then I tried building using 5.0.302 SDK and that succeeded as well, so I tried in a CI pipeline and that also published ok.

**Comment by Samuel12321** [comment]
It's been happening ever since I tried targeting arm64, (Prior to the 23rd).

Unfortunately setting self-contained = false is not an option for us as we need to use self-contained.

There are still a lot of bugs in vs for people trying to publish for arm64.

**Comment by vatsan-madhavan** [comment]
This happened to me again yesterday. I sent MSBuild logs by email to @marcpopMSFT since they’re not publicly shareable - hope they’re helpful in investigating the problem further.

**Comment by marcpopMSFT** [comment]
@vatsan-madhavan We have a bug in 5.0.400 where the work to enable the WindowsDesktop arm64 support that was checked into 5.0.3xx didn't make it into the 5.0.400 release. I'm working on updating our release notes for it and the issue is already resolved and planned for 5.0.401.

https://github.com/dotnet/installer/issues/11481 <-- related bug
https://github.com/dotnet/installer/pull/11458 <-- Fixed issue

**Comment by vatsan-madhavan** [comment]
Building with 5.0.303 seems to work!

**Comment by AArnott** [comment]
@wli3  Why was this closed? It still repros on the 6.0.100 SDK. I opened #22565 to track.
